xarray/core/indexing.py

```diff
@@ -432,16 +447,40 @@ def __init__(self, key):
         super().__init__(new_key)
 
 
-class ExplicitlyIndexed:
+T_ExplicitlyIndexed = TypeVar("T_ExplicitlyIndexed", bound="ExplicitlyIndexed")
+
+
+class ExplicitlyIndexed(metaclass=ABCMeta):
     """Mixin to mark support for Indexer subclasses in indexing."""
 
     __slots__ = ()
 
+    @abstractmethod
+    def __array__(self, dtype: DTypeLike = None) -> np.ndarray:
+        ...
+
+    @abstractmethod
+    def __getitem__(
+        self: T_ExplicitlyIndexed, key: Any
+    ) -> T_ExplicitlyIndexed | NumpyIndexingAdapter | np.ndarray | np.datetime64 | np.timedelta64:
+        ...
+        # TODO: fix the return type? Something more generic?
```

Comment:
Can't it also return something like a `str`, assuming the index holds strings? Or am I misunderstanding this?

Reply:
The problem is the `PandasIndexingAdapter` which has some special casing for datetimes.

Anyway it is quite confusing to me, when you return an Index and when you return a np.ndarray (or scalar for that matter).

Reply:
Agreee it is confusing. 

I think these annotations are quite valuable though. Can we merge a version of this PR that just documents current behaviour? Then we can work to clean it up in a different PR.